File: lib/challenge.rb
# frozen_string_literal:true

LEGEND = %w[M CM D CD C XC L XL X IX V IV I]
         .zip([1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1])

def arabic_to_roman(num)
  return nil if !(num.is_a? Integer) || num <= 0

  result = ''
  LEGEND.each do |roman, arabic|
    while num >= arabic
      result += roman
      num -= arabic
    end
  end
  result
end

def roman_regex
  /(
  (.)\2{3}               # More than three of any character
  |[^MDCLXVI]                           # Illegal characters
  |I[^XVI]|V[^I]|X[^CLXVI]|L[^XVI]       # Illegal sequences
  |C[^MDCLXVI]|D[^MCLXVI]|M[^MDCLXVI]
  )/x
end

def invalid_numeral_string?(numeral_string)
  !(numeral_string.is_a? String) ||
    numeral_string == '' ||
    numeral_string.match(roman_regex)
end

# converts the roman numerals [I, V, X] into their numeric values [1, 5, 10]
def nums_from(numeral_string)
  numeral_string.chars.map { |char| LEGEND.find { |pair| pair[0] == char }[1] }
end

def roman_to_arabic(numeral_string)
  return nil if invalid_numeral_string?(numeral_string)

  nums = nums_from(numeral_string)
  nums.map.with_index do |num, i|
    num * (i < (nums.length - 1) && num < nums[i + 1] ? -1 : 1)
  end.reduce(:+)
end

File: lib/challenge.js
'use strict'

const conversions = [
  [1000, 'M'],
  [900, 'CM'],
  [500, 'D'],
  [400, 'CD'],
  [100, 'C'],
  [90, 'XC'],
  [50, 'L'],
  [40, 'XL'],
  [10, 'X'],
  [9, 'IX'],
  [5, 'V'],
  [4, 'IV'],
  [1, 'I']
]

const arabicToRoman = (num) => {
  if (num !== parseInt(num) || num < 1) {
    return null
  }

  return conversions.map((pair) => {
    const arabic = pair[0]
    const roman = pair[1]
    const str = roman.repeat(Math.floor(num / arabic))
    num %= arabic
    return str
  }).join('')
}

const romanToArabic = (numeralString) => {
  console.log(numeralString)
  if (!(typeof numeralString === 'string' || numeralString instanceof String) || // Not a string
    numeralString === '' || // Empty string
    numeralString.match(/(.)\1{3}/) || // More than three of any character
    numeralString.match(/[^MDCLXVI]/) || // Illegal characters
    numeralString.match(/(I[^XVI]|V[^I]|X[^CLXVI]|L[^XVI])/) || // Illegal sequences
    numeralString.match(/(C[^MDCLXVI]|D[^MCLXVI]|M[^MDCLXVI])/)) {
    return null
  }

  return numeralString
    .split('')
    .map((char) => conversions.find((pair) => pair[1] === char)[0])
    .map((num, i, arr) => {
      return num * (((i < arr.length - 1) && (arr[i] < arr[i + 1])) ? -1 : 1)
    })
    .reduce((a, b) => a + b)
}

module.exports = {
  arabicToRoman,
  romanToArabic
}
